### PR TITLE
Upload component during the create flow

### PR DESCRIPTION
**Pull request by jhadvig**
Adding option to add to project any component that we expose to the users. For this I've created a separate tab in the 'Add to project' page, together with 'From Docker Image'.
Component could be added in a multiple ways:
1. Browse trough the file on users local machine by hitting the "Browse" button right to the input field. 
2. Paste component description into ace editor beneath the input field.
3. Drag and Drop the file from users local machine to the area of the input field or ace editor.
JSON and YAML format are supported.

Also upon componet creation a check is being done to find out whether a component with the same name doesn't exist in current project. If it exists it will be replaced** with the new one. If it doesn't it will be created.
** The way how replacing is working is that we have to delete the old component and create the new one instead. The old component cant be updated with the new one since the new one is missing data like timestamps, revision number, generation , ...
This is a bit unfortunate approach but wasnt sure if extending data of the old component with the data from the new one was better since by extending the fields from the old component would be preserved.

Had to update the oscFileInput scope with an `fileExtension` and `dropArea` variable. Originally wanted to have whole filename, but realize that more valid information then filename(which could be anything) is the file extention. The `dropArea` variable points to a area which should respond to Drag'n'Drop action.

Also for me unknown reason for me when redirecting upon template creation to fromTemplate page and upon imageStream creation to fromImage page  with `$location.path` (create.html#206 && create.html#210) I'm redirected to project list. The path is though right cause when I replace it with `location.assign` method (which is not what we want to use) Im getting the desired redirection.

bindata.go is missing, will provide after review.

Actual screenshot:
screenshot
screenshot-1

@jwforres @spadgett PTAL

### CODE REVIEW

**Comment by jwforres** [comment]
@jhadvig did you try using PATCH instead of POST for the replace?  i'm not sure with the CLI does today for the replace @fabianofranz ?

**Comment by liggitt** [comment]
triggering off the file extension seems fragile to me. can we not do content detection?

**Comment by jwforres** [comment]
@jhadvig you are already doing content detection if the fileExtension is not there (I assume for the direct paste into the editor case?), why not always do content detection?  seems like an unnecessary extra step to validate that the file they upload with extension yml actually parses as YAML.  I think what you are doing with attempting the JSON parse first and then the YAML is probably the right order, since the JSON parser is native to the browser and so hopefully its been optimized.

**Comment by jwforres** [comment]
the Cancel button doesn't seem to be working

**Comment by jwforres** [comment]
Would it be possible to change the ace editor's mode based on what content you have populated it with?  @spadgett do you know if you can change the editor's mode after its created?

looking at JSON when in YAML mode isn't great

**Comment by jwforres** [comment]
I used examples/ha/openshift-ha.yaml and got this:

empty_reason

**Comment by jhadvig** [comment]
@jwforres yes I guess we can skip the file extension check and just do the content detection.
Regarding the ace editor mode, it looks like its possible. Will try to change in upon the aceChanged event and set based on content detection.

**Comment by jhadvig** [comment]
@jwforres I didnt try the PATCH cause I don't think its the right approach/action, cause some fields from the old resource could be left and unwanted, after patching with the new one.
@fabianofranz could you please share info about what CLI does today for the replace action. Thanks !

**Comment by fabianofranz** [comment]
@jhadvig @jwforres `oc replace` does a PUT to update the object, unless it's explicitly set to force with `oc replace --force` then in deletes and creates again (DELETE/POST). `oc patch` does a PATCH, but it's targeted to specifically update some fields instead of the entire object, e.g. `oc patch ... -p '{"spec":{"unschedulable":true}}'`.

File: assets/app/scripts/controllers/create.js [hunk 1]
```diff
@@ -94,6 +100,132 @@ angular.module('openshiftConsole')
...
+      delete $scope.alerts['parsing'];
+      var resource;
+      if ($scope.editor.fileExtension === "yaml") {
+        try {
+          resource = YAML.parse($scope.editor.content);
+        } catch (e) {
+          $scope.alerts['parsing'] = {
+            type: "error",
+            message: "An error occurred during parsing editors content.",
+            details: "Reason: " + $filter('getErrorDetails')(e)
```
Comment (jwforres) [inline assets/app/scripts/controllers/create.js]: getErrorDetails was written to get errors from API responses, the error here is a parse exception, i think you want to take a look at what @spadgett does in the YAML editor as far as handling these exceptions, cause you care about the snippet that the parse error occurred on

File: assets/app/scripts/controllers/create.js [hunk 2]
```diff
@@ -94,6 +100,132 @@ angular.module('openshiftConsole')
...
+          return;
+        }
+      } else if ($scope.editor.fileExtension === "json") {
+        try {
+          resource = JSON.parse($scope.editor.content);
+        } catch (e) {
+          $scope.alerts['parsing'] = {
+            type: "error",
+            message: "An error occurred during parsing editors content.",
+            details: "Reason: " + $filter('getErrorDetails')(e)
```
Comment (jwforres) [inline assets/app/scripts/controllers/create.js]: same thing except a JSON parse exception is probably going to be a different format than the yaml parse exception, just try putting in bad JSON and see what the exception looks like

**Comment by jwforres** [comment]
I think there are two cases that oc create handles that aren't being correctly handled here:
1) kind List, when you see Kind List you are supposed to create each of the things in the List
2) kind SomeOtherKindList, like ImageStreamList, same thing iterate over it and create the things in the list

**Comment by liggitt** [comment]
I think the pattern is

* `kind` = `List` or `*List`
* `items` = Array

**Comment by jhadvig** [comment]
@jwforres I've updated the PR. Added comments to the functions, that should clarify the purpose.
Regarding https://github.com/openshift/origin/pull/8173#issuecomment-199912739 the problem is that we dont have that one it in proper yaml format. Try putting it into http://yamltojson.com/ and convert. It will fail.
PTAL

**Comment by jwforres** [comment]
In the case where you specify a template, I don't like that we actually add the template itself in the project without asking you.  Maybe when someone selects a template we should show something like:

You are adding a template, you can: 
- [x] Process the template - this will create an instance of the things defined in the template, you will have an opportunity to fill out any parameters the template defines
- [ ] Add template - this will make the template available to others who have access to this project
 
@spadgett thoughts?

**Comment by jwforres** [comment]
and by default you would have process checked and add unchecked, since i think that is going to be more common

**Comment by liggitt** [comment]
> and by default you would have process checked and add unchecked, since i think that is going to be more common

+1

**Comment by spadgett** [comment]
@jwforres Sounds good.

**Comment by jwforres** [comment]
@jhadvig also can you re-open this PR against jwforres:console_1.3 branch?  There are enough changes in the branch at this point that testing this PR which is against origin may not catch everything.

**Comment by jwforres** [comment]
also dont include bindata.go in the commit when you open it against console_1.3

**Comment by jhadvig** [comment]
@jwforres @spadgett I've prepared couple of proposals* for the mentioned issue. All of them will pop up a modal that is notifying user that he is adding a Template and that he has to pick up an action. I've picked a modal cause I didnt come up with a better way how to model the picking the next action upon the Template creation (didnt like the idea of scanning the ace editor after every change and dynamically add bunch of buttons, like Save, Process, etc..., or checkboxes).
*following comments present the proposals

**Comment by jhadvig** [comment]
**Proposal 1**
screenshot-7

**Comment by jhadvig** [comment]
**Proposal 2**
screenshot-8
screenshot-9

**Comment by jhadvig** [comment]
**Proposal 3**
In this one I've also added third option and thats *Save and Process*.
Under the selectbox will be the description of the action.
screenshot-3
screenshot-4

**Comment by jhadvig** [comment]
@jwforres @spadgett 
Also the *Add* option should be check by default I think. The reason is that I as a user will be pretty upset when I will be writing the template from scratch into the ace editor and something will go wrong on the process page and my template wont be saved. Thoughts ?  
Thats why I also came up the a standalone option that combines those two(in case we will go with radio or selectbox).

**Comment by spadgett** [comment]
@jhadvig I like proposal 1. Maybe use `class="help-block"` below the checkbox like we use in other places.

**Comment by spadgett** [comment]
@jhadvig Can you rebase your work on https://github.com/jwforres/origin/tree/console_1.3? We've been adding 1.3 content there.

File: assets/app/scripts/controllers/create.js [hunk 3]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+        JSON.parse($scope.editor.content);
+        $scope.editor.session.setMode("ace/mode/json");
+      } catch (e) {
+        try {
+          YAML.parse($scope.editor.content);
+          $scope.editor.session.setMode("ace/mode/yaml");
+        } catch (e) {
+          return;
+        }
+      }
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: We should debounce these checks so we're not parsing and reparsing on every keystroke.

https://lodash.com/docs#debounce

File: assets/app/scripts/controllers/create.js [hunk 4]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
       $scope.filter.tag = tag;
     };
 
+    $scope.aceLoaded = function(editor) {
+      $scope.editor.session = editor.getSession();
+      $scope.editor.session.setOption('tabSize', 2);
+      $scope.editor.session.setOption('useSoftTabs', true);
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: Since this isn't used in the view anywhere, let's make it a local var to the controller.

```javascript
var editorSession;
$scope.aceLoaded = function(editor) {
      editorSession = editor.getSession();

...
```

File: assets/app/scripts/controllers/create.js [hunk 5]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+      });
+    };
+
+    // Check if the editor isn't empty to disable the 'Add' button. Also check in what format the input is in (JSON/YAML) and change
+    // the editor accordingly.
+    $scope.aceChanged = function() {
+      $scope.editor.empty = false;
+      if (!$scope.editor.content) {
+        $scope.editor.empty = true;
+      }
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: I'd just check `!editor.content` in the view and remove `editor.empty`.

File: assets/app/scripts/controllers/create.js [hunk 6]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+        resource = JSON.parse($scope.editor.content);
+      } catch (e) {
+        try {
+          resource = YAML.parse($scope.editor.content);
+        } catch (e) {
+          $scope.alerts['parsing'] = {
+            type: "error",
+            message: "An error occurred during parsing editors content.",
+            details: "Reason: " + e.message
+          };
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: Would be nice to give additional error details like

https://github.com/openshift/origin/blob/master/assets/app/scripts/controllers/modals/editModal.js#L69
https://github.com/openshift/origin/blob/master/assets/app/views/modals/edit-resource.html#L14-L16

File: assets/app/scripts/controllers/create.js [hunk 7]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+          return;
+        }
+      }
+
+      $scope.resourceKind = resource.kind;
+      
+      if ($scope.resourceKind.endsWith("List")) {
+        $scope.resourceList = resource.items;
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]:
```
$scope.resourceName = '';
```

?

File: assets/app/scripts/controllers/create.js [hunk 8]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+        $scope.resourceList = [resource];
+        $scope.resourceName = resource.metadata.name;
+      }
+
+      $scope.updateResources = [];
+      $scope.createResources = [];
+
+      var itemsProcessed = 0;
+      angular.forEach($scope.resourceList, function(item) {
+        var itemKind = item.kind;
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: Show an error if kind if missing.

File: assets/app/scripts/controllers/create.js [hunk 9]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+        var itemName = item.metadata.name;
+
+        if (item.metadata.namespace && item.metadata.namespace !== $scope.projectName) {
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: These lines will result in a runtime error if metadata is missing entirely.

File: assets/app/scripts/controllers/create.js [hunk 10]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+          $scope.alerts['create'] = {
+            type: "error",
+            message: "Can't create component in different projects.",
+            details: itemKind + " " + itemName + " can't be created in namespace " + item.metadata.namespace
+          };
+          return;
+        }
+
+        // Check if the resource already exists. If it does, replace it spec with the new one.
+        DataService.get(APIService.kindToResource(itemKind), itemName, {namespace: $scope.projectName}, {errorNotification: false}).then(
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: Better to move this code somewhere inside `ProjectsService.get()` and use `context` from the callback.

File: assets/app/scripts/controllers/create.js [hunk 11]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+      $scope.createdAndUpdateResources++;
+      if ($scope.createdAndUpdateResources === $scope.resourceList.length) {
+        var subPath;
+        if ($scope.resourceKind === "Template") {
+          subPath =  "create/fromtemplate?name=" + $scope.resourceName + "&namespace=" + encodeURIComponent($scope.projectName);
+        } else {
+          subPath = "overview";
+        }
+        $location.url("project/" + encodeURIComponent($scope.projectName) + "/" + subPath);
+        $scope.$evalAsync();
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: Why is this needed?

File: assets/app/scripts/controllers/create.js [hunk 12]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+            resource.spec = item.spec;
+            $scope.updateResources.push(resource);
+            itemsProcessed++;
+            if (itemsProcessed === $scope.resourceList.length) {
+              if (!_.isEmpty($scope.updateResources)) {
+                openModal();
+              } else {
+                createAndUpdate();
+              }
+            }
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: You should be able to move this code after the `angular.forEach`, then you don't need to track `itemsProcessed`. If you want to track if there were errors, you could set a flag that you check later.

File: assets/app/scripts/controllers/create.js [hunk 13]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+        });
+      });
+    };
+
+    function openModal() {
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: `promptToReplaceResource`?

Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: Or `confirmReplace`?

File: assets/app/scripts/controllers/create.js [hunk 14]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+      var modalInstance = $uibModal.open({
+        animation: true,
+        templateUrl: 'views/modals/update-resource.html',
+        controller: 'UpdateModalController',
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: I'd give the view and controller different names. This sounds like the edit resource modal. Maybe `confirm-replace.html` and `ConfirmReplaceModalController`?

File: assets/app/scripts/controllers/create.js [hunk 15]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+      modalInstance.result.then(function() {
+        createAndUpdate();
+      });
+    }
+
+    // create 
+    function createAndUpdate() {
+      $scope.createdAndUpdateResources = 0;
+
+      if (!_.isEmpty($scope.updateResources)) {
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: The `_.isEmpty` check shouldn't be needed. It's handled by `angular.forEach`.

File: assets/app/scripts/controllers/create.js [hunk 16]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+    }
+
+    // Redirect to appropriate page based on the created/updated resource. If resource is Template redirect to
+    // frotemplate page, else to the project overview.
+    function redirectIfComplete() {
+      $scope.createdAndUpdateResources++;
+      if ($scope.createdAndUpdateResources === $scope.resourceList.length) {
+        var subPath;
+        if ($scope.resourceKind === "Template") {
+          subPath =  "create/fromtemplate?name=" + $scope.resourceName + "&namespace=" + encodeURIComponent($scope.projectName);
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: If creating a list of templates, will this start creating the last template? Probably should go back to the overview.

**Comment by spadgett** [comment]
We should use `TaskList` for create/updating lists of things. Have a look at

https://github.com/openshift/origin/blob/master/assets/app/scripts/controllers/create/createFromImage.js#L214-L258

File: assets/app/scripts/controllers/create.js [hunk 17]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+              redirectIfComplete();
+            });
+        });
+      }  
+    }
+
+    // Redirect to appropriate page based on the created/updated resource. If resource is Template redirect to
+    // frotemplate page, else to the project overview.
+    function redirectIfComplete() {
+      $scope.createdAndUpdateResources++;
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: You should add all the promises to an array and use `$q.all` instead of counting.

https://docs.angularjs.org/api/ng/service/$q#all

File: assets/app/scripts/directives/oscFileInput.js [hunk 18]
```diff
@@ -8,15 +8,25 @@ angular.module('openshiftConsole')
...
       },
       templateUrl: 'views/directives/osc-file-input.html',
       link: function(scope, element){
         scope.helpID = _.uniqueId('help-');
         scope.supportsFileUpload = (window.File && window.FileReader && window.FileList && window.Blob);
         scope.uploadError = false;
+        var dropArea = element;
+        if (scope.dropArea) {
+          dropArea = scope.dropArea;
+        }
```
Comment (spadgett) [inline assets/app/scripts/directives/oscFileInput.js]:
```
var dropArea = scope.dropArea || element;
```

File: assets/app/scripts/directives/oscFileInput.js [hunk 19]
```diff
@@ -31,6 +41,28 @@ angular.module('openshiftConsole')
...
+        $('.btn-file').on('click', function() {
+          delete scope.file;
+        });
+        $(dropArea).on('dragover', function(e) {
+          e.preventDefault();
+          e.stopPropagation();
+        });
+        $(dropArea).on('dragenter', function(e) {
+            e.preventDefault();
+            e.stopPropagation();
```
Comment (spadgett) [inline assets/app/scripts/directives/oscFileInput.js]: Indentation

File: assets/app/scripts/directives/oscFileInput.js (hunk 18 above)
Comment (jwforres) [inline assets/app/scripts/directives/oscFileInput.js]: This reminds me, when I was trying this yesterday... We don't give you any
indication of what the drop target is, I dropped it outside of the editor
area and it replaced the whole page with the contents of the file.

File: assets/app/scripts/directives/oscFileInput.js [hunk 20]
```diff
@@ -31,6 +41,28 @@ angular.module('openshiftConsole')
           };
           reader.readAsText(file);
         });
+        $('.btn-file').on('click', function() {
+          delete scope.file;
```
Comment (spadgett) [inline assets/app/scripts/directives/oscFileInput.js]: Is this triggered when the button is disabled (since it's just a span)? What happens if I cancel the dialog?

File: assets/app/scripts/directives/oscFileInput.js [hunk 21]
```diff
@@ -31,6 +41,28 @@ angular.module('openshiftConsole')
...
+        });
+        $(dropArea).on('drop', function(e) {
+          e.preventDefault();
+          e.stopPropagation();
+          if (e.originalEvent.dataTransfer){
+            if (e.originalEvent.dataTransfer.files.length > 0) {
+              scope.file = e.originalEvent.dataTransfer.files[0];
+              $(element).trigger('change');
+            }
+          }
```
Comment (spadgett) [inline assets/app/scripts/directives/oscFileInput.js]: Maybe

```javascript
var files = _.get(e, 'originalEvent.dataTransferFiles', [])
if (files.length > 0) {
  ...
}
```

File: assets/app/scripts/directives/oscFileInput.js [hunk 22]
```diff
@@ -31,6 +41,28 @@ angular.module('openshiftConsole')
...
+            e.preventDefault();
+            e.stopPropagation();
+        });
+        $(dropArea).on('drop', function(e) {
+          e.preventDefault();
+          e.stopPropagation();
+          if (e.originalEvent.dataTransfer){
+            if (e.originalEvent.dataTransfer.files.length > 0) {
+              scope.file = e.originalEvent.dataTransfer.files[0];
+              $(element).trigger('change');
```
Comment (spadgett) [inline assets/app/scripts/directives/oscFileInput.js]: Is this to trigger the handler above? I'd rather pull that out into its own method and call it directly.

File: assets/app/views/create.html [hunk 23]
```diff
@@ -10,156 +10,199 @@
...
-                          <ul class="uib-dropdown-menu">
-                            <li ng-repeat="tag in browseGeneral" role="menuitem">
-                              <a href="" ng-click="filter.tag = tag">{{tag}}</a>
-                            </li>
-                            <li class="divider"></li>
-                            <li ng-repeat="tag in browseTechnologies" ng-if="tag" role="menuitem">
-                              <a href="" ng-click="filter.tag = tag">{{tag}}</a>
-                            </li>
-                          </ul>
+                <h1>Add Components to your Project</h1>
```
Comment (spadgett) [inline assets/app/views/create.html]: Some push back on the new wording in #7961 :/

File: assets/app/views/create.html [hunk 24]
```diff
@@ -10,156 +10,199 @@
...
+
+                <uib-tabset justified="true">
```
Comment (spadgett) [inline assets/app/views/create.html]: Did we decide if we want justified or not?

File: assets/app/views/create.html [hunk 25]
```diff
@@ -10,156 +10,199 @@
...
+                              showPrintMargin: false
+                            }
+                          }" ng-model="editor.content" class="editor ace_editor-bordered yaml-mode" id="add_component_editor" required></div>
+                          <div class="buttons gutter-top-bottom">
+                            <button
+                              type="submit"
+                              ng-click="create()"
+                              ng-disabled="editor.empty"
+                              class="btn btn-primary btn-lg">
+                              Add
```
Comment (spadgett) [inline assets/app/views/create.html]: Should probably be "Create" to match `oc create`. Also per Patternfly guidelines:

https://www.patternfly.org/styles/terminology-and-wording/#action-labels

File: assets/app/views/create.html [hunk 26]
```diff
@@ -10,156 +10,199 @@
...
+                        </form>
+                      </div>
                     </div>
-                  </div>
-                </div>
+                  </uib-tab>
+
+                  <uib-tab active="selectedTab.fromDockerImage">
+                    <uib-tab-heading>From Docker Image</uib-tab-heading>
+                  </uib-tab>
```
Comment (spadgett) [inline assets/app/views/create.html]: Let's remove the last tab from your PR. I'll add it in mine.

File: assets/app/views/modals/update-resource.html [hunk 27]
```diff
@@ -0,0 +1,26 @@
+<div class="modal-resource-delete">
+  <div class="modal-body">
+    <div ng-if="resourceList.length === 1">
+      <h1>{{resourceKind | humanize}} '<strong>{{resourceName}}</strong>' already exists</h1>
```
Comment (spadgett) [inline assets/app/views/modals/update-resource.html]: This filter was renamed `humanizeKind`.

File: assets/app/views/modals/update-resource.html [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+      <p>Do you want to replace the current '<strong>{{resourceName}}</strong>' {{resourceKind | humanize}} with the submitted one?</p>
+    </div>
+    
+    <div ng-if="resourceList.length > 1">
+      <h1>Some of {{resourceKind | humanize}} items already exists</h1>
```
Comment (spadgett) [inline assets/app/views/modals/update-resource.html]: What if there are multiple resources with different kinds? I'd just say "Some items already exist"

File: assets/app/views/modals/update-resource.html (hunk 27 above)
Comment (spadgett) [inline assets/app/views/modals/update-resource.html]: For when you rebase.

File: assets/app/views/modals/update-resource.html [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+      <p>Following components already exist:</p>
```
Comment (spadgett) [inline assets/app/views/modals/update-resource.html]: "The following items already exist:"

?

File: assets/app/views/modals/update-resource.html [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+      <dl class="dl-horizontal">
+        <div ng-repeat="resource in updateResources">
+          <dt>{{resource.kind}}</dt>
+          <dd>{{resource.metadata.name}}</dd>
+        </div>
```
Comment (spadgett) [inline assets/app/views/modals/update-resource.html]: I think this will work:

```html
<dl class="dl-horizontal">
  <dt ng-repeat-start="resource in updateResources">...</dt>
  <dd ng-repeat-end>...</dd>
</dl>
```

File: assets/app/scripts/controllers/create.js [hunk 31]
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
...
+      $scope.editor.session.setOption('tabSize', 2);
+      $scope.editor.session.setOption('useSoftTabs', true);
+      editor.setDragDelay = 0;
+      editor.$blockScrolling = Infinity;
+
+      $('.from-file .editor').animate({
+        height: Math.floor(window.innerHeight * 0.50) + 'px'
+      }, 30, function() {
+        editor.resize();
+      });
```
Comment (spadgett) [inline assets/app/scripts/controllers/create.js]: Should we handle window resizes?

**Comment by spadgett** [comment]
We should think about breaking up the create controller because it's getting big. Maybe create from file is a directive included in the page.

File: assets/app/views/create.html (hunk 24 above)
Comment (jhadvig) [inline assets/app/views/create.html]: Well there was a discussion and I know that @rhamilto suggested to not have it justified. But then @ajacobs21e supplied the wireframes for the new create page and there it was justified, so I've figured out I will leave it justified and lets discuss it on the PR, here :)

File: assets/app/views/modals/update-resource.html (hunk 27 above)
Comment (jhadvig) [inline assets/app/views/modals/update-resource.html]: :+1:

File: assets/app/views/modals/update-resource.html (hunk 28 above)
Comment (jhadvig) [inline assets/app/views/modals/update-resource.html]: There can be only one top level `Kind`, in this particular case it will be List. Listed will be only those items from the List that already exist, whatever type they are.

File: assets/app/views/modals/update-resource.html (hunk 29 above)
Comment (jhadvig) [inline assets/app/views/modals/update-resource.html]: Hmmm, will merge the h1 and p into one header saying:
"Following List items already exists"

File: assets/app/scripts/directives/oscFileInput.js (hunk 22 above)
Comment (jhadvig) [inline assets/app/scripts/directives/oscFileInput.js]: You mean that the `$(element).trigger('change');` will be in a standalone method,which we can then call to trigger?

File: assets/app/scripts/directives/oscFileInput.js (hunk 21 above)
Comment (jhadvig) [inline assets/app/scripts/directives/oscFileInput.js]: :+1:

File: assets/app/scripts/controllers/create.js (hunk 8 above)
Comment (jhadvig) [inline assets/app/scripts/controllers/create.js]: Well I though about adding and error but I wasn't sure if we wanna have the validation done client side. Then we would have to check for all the fields that have to be filled for each resources kind. Do we want it ?

Comment (jhadvig) [inline assets/app/scripts/controllers/create.js]: Sorry I take it back, I've realized that we need it.

File: assets/app/views/modals/update-resource.html (hunk 28 above)
Comment (spadgett) [inline assets/app/views/modals/update-resource.html]: So it says "Some of List items already exists"? In that case, I'd still just say "Some items already exist"

Comment (jhadvig) [inline assets/app/views/modals/update-resource.html]: No I meant, it will say: `Following components already exist:`

Comment (spadgett) [inline assets/app/views/modals/update-resource.html]: Let's say items for now until we get agreement on terminology.

Comment (jhadvig) [inline assets/app/views/modals/update-resource.html]: :+1:

File: assets/app/scripts/controllers/create.js (hunk 16 above)
Comment (jhadvig) [inline assets/app/scripts/controllers/create.js]: `$scope.resourceKind` is holding the top level resource Kind. So in the case of TemplateList it will be redirected to overview

**Comment by jhadvig** [comment]
@spadgett rebased && added the bindata

File: assets/app/scripts/controllers/newfromtemplate.js [hunk 32]
```diff
@@ -188,17 +196,38 @@ angular.module('openshiftConsole')
...
+          // In case the template can be loaded from 'CachedTemaplteService', show an alert and disable "Create" button.
+          if (_.isEmpty($scope.template)) {
+            $scope.alerts["process"] =
+              {
+                type: "error",
+                message: "Unable to load the Template.",
+                details: "Not found in cache."
+              };
+            $scope.templateForm.$setValidity("",false);
+            return;
```
Comment (spadgett) [inline assets/app/scripts/controllers/newfromtemplate.js]: We should navigate to the error page here, or improve the error message. It currently looks like this.

openshift_web_console

**Comment by spadgett** [comment]
Seeing some runtime errors on add to project after picking a source builder:

openshift_web_console

File: assets/app/scripts/directives/fromFile.js [hunk 33]
```diff
@@ -0,0 +1,397 @@
...
+            $scope.isList = true;
+            $scope.resourceList = resource.items;
+            $scope.resourceName = '';
+          } else {
+            $scope.resourceList = [resource];
+            $scope.resourceName = resource.metadata.name;
+            if ($scope.resourceKind === "Template") {
+              $scope.templateOptions = {
+                process: true,
+                add: true
```
Comment (spadgett) [inline assets/app/scripts/directives/fromFile.js]: Missed this on earlier review, but @jwforres and @liggitt suggested add be false by default.

File: assets/app/scripts/controllers/newfromtemplate.js (hunk 32 above)
Comment (jhadvig) [inline assets/app/scripts/controllers/newfromtemplate.js]: Yeah, redirecting to error page would be better

**Comment by jhadvig** [comment]
@spadgett all three fixed. Thanks !

**Comment by jhadvig** [comment]
The thing I wondering about is if we dont wanna have the 'From catalog' and 'From file' tab headers justified. Or not justifying was the final decision ? 
@spadgett 
cc'ing @jwforres @rhamilto

**Comment by spadgett** [comment]
> The thing I wondering about is if we dont wanna have the 'From catalog' and 'From file' tab headers justified. 

I prefer unjustified for what it's worth.

**Comment by spadgett** [comment]
@jhadvig The projects dropdown stops working for me once I drag a file. Do you see this as well?

**Comment by spadgett** [comment]
Add to Project -> From File -> Drop a file -> Click the project name in the breadcrumb -> Click the projects downdrop

File: assets/app/views/create.html (hunk 23 above)
Comment (spadgett) [inline assets/app/views/create.html]: Let's use the header "Add to Project" until we have API approval to make a change (see #7961)

**Comment by jhadvig** [comment]
@spadgett yes I can, but its really weird !?! Will take a look whats causing the problem

**Comment by jhadvig** [comment]
hmm ... I can even reproduce the issue when redirecting to 'Add to project' and then go back to the overview

File: assets/app/scripts/directives/oscFileInput.js [hunk 34]
```diff
@@ -32,6 +141,19 @@ angular.module('openshiftConsole')
...
+          $('.drag-and-drop-zone').removeClass("show-drag-and-drop-zone highlight-drag-and-drop-zone");
+        }
+
+        function removeDropZoneElements() {
+          $(dropZoneSelectorName).remove();
+        }
+
+        scope.$on('$destroy', function(){
+          $(dropZoneSelectorName).off();
+          $(document).off('drop').off('dragenter').off('dragover').off('dragleave');
```
Comment (spadgett) [inline assets/app/scripts/directives/oscFileInput.js]: Can you give these namespaces so we don't accidentally remove other drag/drop event handlers that might be registered? Maybe `"drop.oscFileInput"` when you call `on()`. Then `$(document).off('.oscFileInput')` should be enough to unregister all the different event types (untesed).

http://api.jquery.com/on/
http://api.jquery.com/off/

**Comment by jhadvig** [comment]
@spadgett thanks so much for the great suggestion ! I've updated the PR

**Comment by spadgett** [comment]
LGTM, @jwforres suggested a fork AMI so QE can test IE

**Comment by jhadvig** [comment]
@spadgett ok will create one right away

**Comment by spadgett** [comment]
[merge]

**Comment by jhadvig** [comment]
@spadgett after your PR get merger I had to rebase my (not fully sure why though...)
But its rebased now :)

**Comment by spadgett** [comment]
[merge]

**Comment by jhadvig** [comment]
@spadgett thanks so much for assistance on this !!!
:joy:
